// remotion/dist/cjs/canvas-effects/use-effect-chain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useEffectChain = void 0;
const react_1 = require("react");
const use_current_frame_js_1 = require("../use-current-frame.js");
const use_delay_render_js_1 = require("../use-delay-render.js");
const canvas_pool_js_1 = require("./canvas-pool.js");
const effect_internals_js_1 = require("./effect-internals.js");
const gpu_device_js_1 = require("./gpu-device.js");
const ensureSetup = (state, def, target) => {
    const cached = state.setupCache.get(def);
    if (cached) {
        return cached;
    }
    const promise = Promise.resolve(def.setup(target));
    state.setupCache.set(def, promise);
    state.cleanupRegistry.push({ definition: def, statePromise: promise });
    return promise;
};
// `useEffectChain` is the shared per-frame runtime that powers source
// components like `<Solid>` and `<HtmlInCanvas>`. Source components are
// responsible for producing a `CanvasImageSource` via the `source` callback;
// the runtime owns the rest of the pipeline (scratch canvases, ping-pong,
// cross-backend transfer via `createImageBitmap`, single `delayRender` per
// frame).
const useEffectChain = ({ source, effects, width, height, pixelRatio = 1, output, sourceDeps, }) => {
    const frame = (0, use_current_frame_js_1.useCurrentFrame)();
    const { delayRender, continueRender, cancelRender } = (0, use_delay_render_js_1.useDelayRender)();
    const stateRef = (0, react_1.useRef)(null);
    const sizeRef = (0, react_1.useRef)(null);
    // Reset pool if dimensions changed.
    if (!sizeRef.current ||
        sizeRef.current.width !== width ||
        sizeRef.current.height !== height) {
        stateRef.current = {
            pool: new canvas_pool_js_1.CanvasPool(width, height),
            setupCache: new WeakMap(),
            cleanupRegistry: [],
        };
        sizeRef.current = { width, height };
    }
    const flattened = (0, react_1.useMemo)(() => (0, effect_internals_js_1.flattenEffects)(effects), [effects]);
    const runs = (0, react_1.useMemo)(() => (0, effect_internals_js_1.groupByBackend)(flattened), [flattened]);
    (0, react_1.useEffect)(() => {
        const state = stateRef.current;
        if (!state || !output) {
            return;
        }
        const handle = delayRender(`Canvas effect chain (frame ${frame})`);
        let cancelled = false;
        let resolved = false;
        const finish = () => {
            if (resolved) {
                return;
            }
            resolved = true;
            continueRender(handle);
        };
        const fail = (err) => {
            if (resolved) {
                return;
            }
            resolved = true;
            cancelRender(err);
        };
        (async () => {
            var _a;
            var _b;
            try {
                const sourceImage = await source();
                if (cancelled) {
                    return;
                }
                if (!sourceImage) {
                    finish();
                    return;
                }
                let currentImage = sourceImage;
                let lastTarget = null;
                if (runs.length === 0) {
                    // No effects: blit source directly into output.
                    const ctx = output.getContext('2d');
                    if (!ctx) {
                        throw new Error('Failed to acquire 2D context for output canvas');
                    }
                    ctx.clearRect(0, 0, width, height);
                    ctx.drawImage(currentImage, 0, 0, width, height);
                    finish();
                    return;
                }
                let needsGpuDevice = false;
                for (const run of runs) {
                    if (run.backend === 'webgpu') {
                        needsGpuDevice = true;
                        break;
                    }
                }
                const gpuDevice = needsGpuDevice ? await (0, gpu_device_js_1.getGpuDevice)() : null;
                if (cancelled) {
                    return;
                }
                for (let runIndex = 0; runIndex < runs.length; runIndex++) {
                    const run = runs[runIndex];
                    const [a, b] = state.pool.getPair(run.backend);
                    let dst = a;
                    for (const eff of run.effects) {
                        const def = eff.definition;
                        const setupState = await ensureSetup(state, def, dst);
                        if (cancelled) {
                            return;
                        }
                        await def.apply({
                            source: currentImage,
                            target: dst,
                            state: setupState,
                            params: eff.params,
                            frame,
                            width,
                            height,
                            pixelRatio,
                            gpuDevice,
                        });
                        if (cancelled) {
                            return;
                        }
                        currentImage = dst;
                        dst = dst === a ? b : a;
                    }
                    lastTarget = (_b = currentImage) !== null && _b !== void 0 ? _b : lastTarget;
                    // If another run follows with a different backend, transfer via
                    // `createImageBitmap` so the next backend can sample from a
                    // neutral GPU-resident handle.
                    const nextRun = runs[runIndex + 1];
                    if (nextRun && nextRun.backend !== run.backend && lastTarget) {
                        const bitmap = await createImageBitmap(lastTarget);
                        if (cancelled) {
                            (_a = bitmap.close) === null || _a === void 0 ? void 0 : _a.call(bitmap);
                            return;
                        }
                        currentImage = bitmap;
                    }
                }
                if (!lastTarget) {
                    finish();
                    return;
                }
                const outCtx = output.getContext('2d');
                if (!outCtx) {
                    throw new Error('Failed to acquire 2D context for output canvas');
                }
                outCtx.clearRect(0, 0, width, height);
                outCtx.drawImage(lastTarget, 0, 0, width, height);
                finish();
            }
            catch (err) {
                if (!cancelled) {
                    fail(err);
                }
            }
        })();
        return () => {
            cancelled = true;
            if (!resolved) {
                continueRender(handle);
            }
        };
        /* eslint-disable react-hooks/exhaustive-deps */
    }, [
        frame,
        width,
        height,
        pixelRatio,
        output,
        runs,
        delayRender,
        continueRender,
        cancelRender,
        ...sourceDeps,
    ]);
    /* eslint-enable react-hooks/exhaustive-deps */
    // Cleanup on unmount: invoke `cleanup` for every cached state.
    (0, react_1.useEffect)(() => {
        return () => {
            const state = stateRef.current;
            if (!state) {
                return;
            }
            for (const entry of state.cleanupRegistry) {
                entry.statePromise.then((s) => {
                    var _a, _b;
                    return (_b = (_a = entry.definition).cleanup) === null || _b === void 0 ? void 0 : _b.call(_a, s);
                }, () => undefined);
            }
        };
    }, []);
};
exports.useEffectChain = useEffectChain;
